File: sdks/go/docflow/storage/storage.go
// Package storage provides storage backends for DocFlow.
package storage

import (
	"io"
)

// Storage defines the interface for storing and retrieving files.
type Storage interface {
	// Save stores data at the given path.
	Save(path string, data []byte) error

	// SaveReader stores data from a reader at the given path.
	SaveReader(path string, reader io.Reader) error

	// Load retrieves data from the given path.
	Load(path string) ([]byte, error)

	// LoadReader returns a reader for the data at the given path.
	LoadReader(path string) (io.ReadCloser, error)

	// Delete removes the file at the given path.
	Delete(path string) error

	// Exists checks if a file exists at the given path.
	Exists(path string) (bool, error)

	// List returns all files in the given directory.
	List(dir string) ([]string, error)

	// GetURL returns a URL for accessing the file (if supported).
	// Returns empty string if not supported.
	GetURL(path string) string
}

// StorageType represents the type of storage backend.
type StorageType string

const (
	StorageTypeLocal StorageType = "local"
	StorageTypeS3    StorageType = "s3"
	StorageTypeGCS   StorageType = "gcs"
	StorageTypeAzure StorageType = "azure"
)

File: sdks/java/src/main/java/com/docflow/storage/Storage.java
package com.docflow.storage;

import java.io.IOException;
import java.io.InputStream;
import java.util.List;
import java.util.Optional;

/**
 * Interface for file storage backends.
 */
public interface Storage {

    /**
     * Save data to the given path.
     *
     * @param path Relative path where to save the file.
     * @param data Binary data to save.
     * @throws IOException If saving fails.
     */
    void save(String path, byte[] data) throws IOException;

    /**
     * Save data from an input stream.
     *
     * @param path        Relative path where to save the file.
     * @param inputStream Stream to read from.
     * @throws IOException If saving fails.
     */
    void save(String path, InputStream inputStream) throws IOException;

    /**
     * Load data from the given path.
     *
     * @param path Relative path to the file.
     * @return Binary file contents.
     * @throws IOException If loading fails or file doesn't exist.
     */
    byte[] load(String path) throws IOException;

    /**
     * Delete file at the given path.
     *
     * @param path Relative path to the file.
     * @throws IOException If deletion fails.
     */
    void delete(String path) throws IOException;

    /**
     * Check if file exists at the given path.
     *
     * @param path Relative path to check.
     * @return true if file exists.
     */
    boolean exists(String path);

    /**
     * List files in the given directory.
     *
     * @param directory Relative path to directory.
     * @return List of filenames in the directory.
     * @throws IOException If listing fails.
     */
    List<String> list(String directory) throws IOException;

    /**
     * Get URL for accessing the file.
     *
     * @param path Relative path to the file.
     * @return URL string if supported, empty otherwise.
     */
    default Optional<String> getUrl(String path) {
        return Optional.empty();
    }
}
